### PR TITLE
[FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88260 [FSDP()][Easy] Make `fully_shard()` only `FULL_SHARD`
* #88235 [FSDP()] Have `fully_shard()` abide by `@contract`!
* #88234 [FSDP()][Easy] Rename `_State` to `_FSDPState`
* #88233 [FSDP()] Rename to `fully_shard()` and move to `_composable/`
* #88232 [FSDP][Easy] Remove unneeded `TrainingState` transition
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87935 [FSDP()][21/N] Refactor and fix `_cast_buffers()`
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* **#87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`**
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward

This PR moves `_fsdp_root_pre_forward()` to `_runtime_utils.py`.

Note: This PR includes a (temporary) fix for `NO_SHARD` + `CPUOffload(offload_params=True)`, where we set `non_blocking=False` when copying the gradient from device to host. It is only included in this PR since the test was **flaky** (but not consistently failing) on this PR , so I needed to fix to unblock land.

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -180,6 +181,43 @@ def _post_forward(
...
+    before any individual module's pre-forward. If this is called on a non-root
+    FSDP instance, then the forward inputs are returned directly.
+    """
+    p_assert(state._is_root is not None, "Expects a root FSDP to have been set")
+    if not state._is_root:
+        return args, kwargs
+    if state.forward_prefetch:
+        handles_keys = []
+        if _is_composable(state):
+            # TODO: This assumes singleton handles keys.
```
Comment (awgu) [inline torch/distributed/fsdp/_runtime_utils.py:201]: Basically, if we ever have more than one `FlatParamHandle` per module, then we need to track which ones are grouped together for the composable code path.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -180,6 +181,43 @@ def _post_forward(
...
+                handles_key = tuple(fsdp_module._handles)
+        for handles_key in handles_keys:
+            state._needs_pre_forward_unshard[handles_key] = True
+    _wait_for_computation_stream(
+        torch.cuda.current_stream(),
+        state._streams["unshard"],
+        state._streams["pre_unshard"],
+    )
+    _clear_grads_if_needed(_all_handles(state))
+    input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:214]: hmm, this part looks a bit different, it used to be:

```python
        input_dtype = (
            self.mixed_precision.param_dtype
            if self._mixed_precision_enabled_for_params()
            else None
        )
```

**Review by mrshenli — APPROVED** [review]
stamp to unblock, please double check the changes on `mixed_precision `

File: torch/distributed/fsdp/_runtime_utils.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:214]: Yup. This is intended and just a simplification. `self._mixed_precision_enabled_for_params()` is exactly just `return self.mixed_precision.param_dtype is not None`.

File: torch/distributed/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -180,6 +181,43 @@ def _post_forward(
...
+    if not state._is_root:
+        return args, kwargs
+    if state.forward_prefetch:
+        handles_keys = []
+        if _is_composable(state):
+            # TODO: This assumes singleton handles keys.
+            handles_keys = [tuple(handle) for handle in state._handles]
+        else:
+            for fsdp_module in state.fsdp_modules(state):
+                handles_key = tuple(fsdp_module._handles)
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/_runtime_utils.py:205]: this seems not to put handles_key into handles_keys list?

Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:205]: Great catch!
